packages/lesswrong/lib/ultraFeed/ultraFeedPostHelpers.ts

```diff
@@ -0,0 +1,170 @@
+    return [];
+  }
+  
+  // Construct the RQL filter string ONLY for the excludedPostIds
+  let exclusionFilterString: string | undefined = undefined;
+  if (excludedPostIds.length > 0) {
+      const exclusionFilter = excludedPostIds.map(id => `"${id}"`).join(',');
+      exclusionFilterString = `'itemId' NOT IN {${exclusionFilter}}`;
+  }
+  
+  const lwAlgoSettings: RecombeeRecommendationArgs = {
+      scenario: scenarioId,
+      filterSettings: currentUser?.frontpageFilterSettings,
+      ...(exclusionFilterString && { filter: exclusionFilterString }),
+  };
+  
+  try {
+      const recommendedResults = await recombeeApi.getRecommendationsForUser( recombeeUser, limit, lwAlgoSettings, context);
+
+      const displayPosts = recommendedResults.map((item): FeedFullPost | null => {
+          if (!item.post?._id) return null;
+          const { post, recommId, scenario, generatedAt } = item;
+          
+          const recommInfo = (recommId && generatedAt) ? {
+              recommId,
+              scenario: scenario || scenarioId,
+              generatedAt,
+          } : undefined;
+          
+          return {
+              post,
+              postMetaInfo: {
+                  sources: [scenario as FeedItemSourceType],
+                  displayStatus: 'expanded',
+                  recommInfo: recommInfo,
+              },
+          };
+      }).filter((p) => !!p);
+
+      return displayPosts;
+
+  } catch (error) {
+      // eslint-disable-next-line no-console
+      console.error(`Error calling getRecommendationsForUser for scenario ${scenarioId}:`, error);
+      return [];
+  }
+}
+
+/**
+ * Fetches the latest posts and maps them directly to FeedFullPost format.
+ */
+export async function getLatestPostsForUltraFeed(
+  context: ResolverContext,
+  limit: number,
+  excludedPostIds: string[] = [],
+): Promise<FeedFullPost[]> {
+  const latestPostsSourceType: FeedItemSourceType = 'hacker-news';
+  const { currentUser } = context;
+  const hiddenPostIds = currentUser?.hiddenPostsMetadata?.map(metadata => metadata.postId) ?? [];
+  const allExcludedIds = [...new Set([...excludedPostIds, ...hiddenPostIds])];
+
+  const thirtyDaysAgo = new Date(new Date().getTime() - (30*24*60*60*1000));
+  const filterSettings: FilterSettings = currentUser?.frontpageFilterSettings ?? getDefaultFilterSettings();
+
+  const postsTerms: PostsViewTerms = {
+    view: "magic",
+    forum: true,
+    limit,
+    filterSettings,
+    after: thirtyDaysAgo,
+    karmaThreshold: 10,
+    ...(allExcludedIds.length > 0 && { notPostIds: allExcludedIds }),
+  };
+
+  const postsQuery = viewTermsToQuery('Posts', postsTerms, undefined, context);
+  
+  try {
+    const posts = await performQueryFromViewParameters(context.Posts, postsTerms, postsQuery);
+    const potentiallyPartialPosts = await accessFilterMultiple(currentUser, 'Posts', posts, context);
+    const accessiblePosts = filterNonnull(potentiallyPartialPosts) as DbPost[];
```

Comment:
this cast should be unnecessary (and also I don't even think the filter is necessary; `accessFilterMultiple` doesn't return arrays with `null` values in it)